### ISSUE TITLE
[RFC] Expand on how to extend classes in documentation and examples

### ISSUE THREAD

**Issue by SeanNaren** [PyTorchLightning/lightning-transformers#174]
## 🚀 Feature

Seems like many use cases involve extending classes, and modifying things.

This can be very confusing as a it involves a lot of steps for a complete workflow, but sometimes you only need to do a few steps (like if you do not care about Hydra).

It might be better to do a use case approach to things:

1. I want to extend the loss and modify the loss function
2. I want to train on my own dataset and my processing requires modification to the standard dataset processing
3. I want to modify the model that we load, modifying the layers or freeze layers
4. I want to train my own model not HF model (like a from scratch `nn.Module` )

If anyone has any other cases they'd like covered/cleared up let me know! I think its important all of these become examples in the documentation and the code supports these cases. It should also be clearer when Hydra can be used on all of these to improve iteration :)
